### PR TITLE
Issue #2766: Restore missing "Page by date" Views pager plugin

### PR DESCRIPTION
**Pull request by Al-Rozhkov**
...and Date fields contextual filter.

### CODE REVIEW

File: core/modules/date/css/date-views.css [hunk 1]
```diff
@@ -1,3 +1,23 @@
+/* Pager plugin css */
+div.date-views-pager {
```
Comment (jenlampton) [inline core/modules/date/css/date-views.css]: Can we remove the `div` from these selectors? The more general we can be here the easier it will be for contrib to override the styles where necessary.

File: core/modules/date/date.theme.inc [hunk 2]
```diff
@@ -474,4 +474,223 @@ function template_preprocess_date_views_filter_form(&$vars) {
...
+      default:
+        // Already on the bottom.
+    }
+  }
+}
+
+/**
+ * Preprocess function for Date pager template.
+ */
+function template_preprocess_date_views_pager(&$vars) {
```
Comment (jenlampton) [inline core/modules/date/date.theme.inc]: All instances of `$vars` should be replaced with `$variables` as per our coding standards. It looks like this code may be left over from Drupal 6.

File: core/modules/date/templates/date-views-pager.tpl.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * @file
+ * Template file for the example display.
+ *
+ * Variables available:
```
Comment (jenlampton) [inline core/modules/date/templates/date-views-pager.tpl.php]: Can we format the `Available variables:` section  to match the rest of core's templates, for example:

```php
 * Available variables:
 * - $plugin: The pager plugin object. This contains the view.
 * - $plugin->view: The view object for this navigation.
 * - $nav_title: The formatted title for this view. In the case of block
 *   views, it will be a link to the full view, otherwise it will
 *   be the formatted name of the year, month, day, or week.
 * - $prev_url: Url for the previous calendar page. The link is created
 *    in the template to make it easier to change the text, add images, etc.
 * - $next_url: Url for the next calendar page. The link is created
 *    in the template to make it easier to change the text, add images, etc.
```

File: core/modules/date/templates/date-views-pager.tpl.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * $prev_options
+ * $next_options
+ *   Query strings and other options for the links that need to
+ *   be used in the l() function, including rel=nofollow.
+ */
+?>
+<?php if (!empty($pager_prefix)) : ?>
+<?php print $pager_prefix; ?>
+<?php endif; ?>
+<div class="date-nav-wrapper clearfix<?php if (!empty($extra_classes)): print $extra_classes; endif; ?>">
```
Comment (jenlampton) [inline core/modules/date/templates/date-views-pager.tpl.php]: These classes should be an array in preprocess, and then printed here using `<?php print implode(' ', $classes); ?>` to match with how we handle classes in Backdrop template files.

File: core/modules/date/views/date_views_argument_handler.inc [hunk 5]
```diff
@@ -10,7 +10,25 @@
 /**
  * Date API argument handler.
  */
+// @codingStandardsIgnoreStart
```
Comment (jenlampton) [inline core/modules/date/views/date_views_argument_handler.inc]: This line should probably be removed.

File: core/modules/date/views/date_views_argument_handler.inc [hunk 6]
```diff
@@ -108,8 +126,10 @@ class date_views_argument_handler extends date_views_argument_handler_simple {
     // There is no easy way to do summary queries on multiple fields, so remove that option.
     unset($defaults['summary']);
 
-    if ($which && !empty($defaults[$which])) {
-      return $defaults[$which];
+    if ($which) {
```
Comment (jenlampton) [inline core/modules/date/views/date_views_argument_handler.inc]: Is there a good reason to separate the comparisons here? Not a biggie, just curious. :)

File: core/modules/date/views/date_views_argument_handler_simple.inc [hunk 7]
```diff
@@ -7,49 +7,10 @@
 /**
  * Date API argument handler.
  */
+// @codingStandardsIgnoreStart
```
Comment (jenlampton) [inline core/modules/date/views/date_views_argument_handler_simple.inc]: another cleanup here.

File: core/modules/date/views/date_views_argument_handler.inc [hunk 8]
```diff
@@ -164,17 +186,18 @@ class date_views_argument_handler extends date_views_argument_handler_simple {
...
         $date_handler->local_timezone = date_get_timezone($field['tz_handling']);
         date_views_set_timezone($date_handler, $this, $field);
-        $query_field[] = array('field' => $field, 'date_handler' => $date_handler);
+        $this->query_fields[] = array('field' => $field, 'date_handler' => $date_handler);
       }
     }
   }
 
 }
+// @codingStandardsIgnoreEnd
```
Comment (jenlampton) [inline core/modules/date/views/date_views_argument_handler.inc]: cleanup.

File: core/modules/date/views/date_views_argument_handler_simple.inc [hunk 9]
```diff
@@ -420,3 +383,4 @@ class date_views_argument_handler_simple extends views_handler_argument_date {
   }
 
 }
+// @codingStandardsIgnoreEnd
```
Comment (jenlampton) [inline core/modules/date/views/date_views_argument_handler_simple.inc]: cleanup

File: core/modules/date/views/date_views_fields.inc [hunk 10]
```diff
@@ -24,6 +24,7 @@ function _date_views_fields($base = 'node') {
 
   // We use fields that provide filter handlers as our universe of possible
   // fields of interest.
+  module_load_include('inc', 'views_ui', 'views_ui.admin');
```
Comment (jenlampton) [inline core/modules/date/views/date_views_fields.inc:27]: I wonder why this was needed. Did things explode without it?

File: core/modules/date/views/date_views_plugin_pager.inc [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+ * we can swap the pager value into the query created
+ * by the date argument and adjust the
+ * $view->date_info values set by the argument accordingly
+ * so the theme will pick up the new information.
+ */
+
+/**
+ * Example plugin to handle paging by month.
+ */
+// @codingStandardsIgnoreStart
```
Comment (jenlampton) [inline core/modules/date/views/date_views_plugin_pager.inc]: cleanup

File: core/modules/date/views/date_views_plugin_pager.inc [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+  }
+
+
+  function render($input) {
+    // This adds all of our template suggestions based upon the view name and display id.
+    $pager_theme = views_theme_functions('date_views_pager', $this->view, $this->display);
+    return theme($pager_theme, array('plugin' => $this, 'input' => $input));
+  }
+}
+// @codingStandardsIgnoreEnd
```
Comment (jenlampton) [inline core/modules/date/views/date_views_plugin_pager.inc]: cleanup

File: core/modules/date/views/date.views.inc [hunk 13]
```diff
@@ -60,7 +54,7 @@ function date_views_views_plugins() {
 /**
  * Implements hook_views_data().
  */
-function date_views_views_data() {
+function date_views_data() {
```
Comment (quicksketch) [inline core/modules/date/views/date.views.inc:57]: Yikes, this looks like a big fix! This might be worth separating into a stand-alone PR.

File: core/modules/date/views/date.views.inc [hunk 14]
```diff
@@ -98,7 +92,7 @@ function date_views_views_data() {
 /**
  * Implements hook_views_data_alter().
  */
-function date_views_views_data_alter(&$data) {
+function date_views_data_alter(&$data) {
```
Comment (quicksketch) [inline core/modules/date/views/date.views.inc:95]: +1

File: core/modules/date/views/date_views_argument_handler_simple.inc [hunk 15]
```diff
@@ -7,49 +7,10 @@
...
-  /**
-   * @var DateTime
-   */
-  protected $min_date;
-
-  /**
-   * @var DateTime
-   */
-  protected $max_date;
-
```
Comment (quicksketch) [inline core/modules/date/views/date_views_argument_handler_simple.inc]: Seems like all these variables should stay defined, unless they actually do not exist.

File: core/modules/date/views/date.views.inc (hunk 13 above)
Comment (jenlampton) [reply core/modules/date/views/date.views.inc:57]: And this one would depend on that one?

File: core/modules/date/views/date_views_argument_handler_simple.inc [hunk 16]
```diff
@@ -315,7 +278,7 @@ class date_views_argument_handler_simple extends views_handler_argument_date {
     $this->base_alias = $this->name_alias = $this->query->add_field(NULL, $formula, $this->field . '_summary');
     $this->query->set_count_field(NULL, $formula, $this->field);
 
-    $this->summary_basics(FALSE);
+    return $this->summary_basics(FALSE);
```
Comment (quicksketch) [inline core/modules/date/views/date_views_argument_handler_simple.inc]: This return seems unnecessary, `summary_basics` doesn't return anything?

File: core/modules/date/views/date_views_fields.inc (hunk 10 above)
Comment (Al-Rozhkov) [reply core/modules/date/views/date_views_fields.inc:27]: Without it I have `Error: Call to undefined function views_fetch_fields()`

File: core/modules/date/views/date_views_plugin_pager.inc [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+ * the original query. The pager query
+ * runs after that, so the pager checks to see
+ * if there is a pager value that needs to be used in the query.
+ * The date argument has identified the placeholders
+ * it used in the query. So if a change is needed,
+ * we can swap the pager value into the query created
+ * by the date argument and adjust the
+ * $view->date_info values set by the argument accordingly
+ * so the theme will pick up the new information.
+ */
```
Comment (quicksketch) [inline core/modules/date/views/date_views_plugin_pager.inc]: This docblock should be formatted to be 80 characters wide, right now it seems random where the lines are wrapping.

This should be moved from `@file` documentation to being the documentation for the `date_views_plugin_pager` class itself.

File: core/modules/date/views/date_views_plugin_pager.inc [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+
+/**
+ * Example plugin to handle paging by month.
+ */
+
+class date_views_plugin_pager extends views_plugin_pager {
```
Comment (quicksketch) [inline core/modules/date/views/date_views_plugin_pager.inc:56]: Let's remove this "Example plugin to handle paging by month" docblock and replace it with the above docblock.

File: core/modules/date/views/date_views_argument_handler_simple.inc [hunk 19]
```diff
@@ -7,47 +7,8 @@
...
-
-  /**
-   * @var DateTime
-   */
-  protected $min_date;
-
-  /**
-   * @var DateTime
-   */
-  protected $max_date;
```
Comment (quicksketch) [inline core/modules/date/views/date_views_argument_handler_simple.inc]: Restore all these property definitions.

**Review by quicksketch — CHANGES_REQUESTED** [review]
I haven't given this a full pass, but minor updates still needed.

**Comment by quicksketch** [comment]
Closing/reopening for a new sandbox.

**Comment by docwilmot** [comment]
@Al-Rozhkov can you rebase this? Was going to check it but there are conflicts.

**Comment by stpaultim** [comment]
@Al-Rozhkov 

I just found a link to this PR for an issue that is already closed. 
https://github.com/backdrop/backdrop-issues/issues/2766

I think we could close this PR.
